### DISCUSSION TITLE
Type Safely Handling Prisma Model Names in Class Constructor

### DISCUSSION THREAD

**Discussion by Lucasnhso** [prisma/prisma#21768] [Q&A]
### Question

I created a BaseRepository class that will serve as the basis for all the other repositories that I will create, for this I need to receive the modelName as a parameter, I am not able to type it, the only Prisma type that I found that could be typed is ```Prisma.ModelName```, but It only works if my model's name is defined starting with a lowercase letter, if it is uppercase it doesn't work, because it would try to access ```prismaClient.User``` for example, but there is only ```prismaClient.user```.

```js
import { PrismaClient, Prisma } from "@prisma/client";

export default class BaseRepository {
  protected model;
  protected prismaClient = new PrismaClient();

  constructor(modelName: Prisma.ModelName) {
    this.model = this.prismaClient[modelName];
    this.findMany = this.findMany.bind(this);
    this.create = this.create.bind(this);
    this.findById = this.findById.bind(this);
    this.update = this.update.bind(this);
    this.delete = this.delete.bind(this);
  }

  public async findMany() {
    return await this.model.findMany();
  }

  public async findById(id: string) {
    return await this.model.findUnique({
      where: {
        id,
      },
    });
  }
  public async create(data: any) {
    return await this.model.create({
      data: {
        ...data,
      },
    });
  }

  public async update(id: string, data: any) {
    return await this.model.update({
      where: { id },
      data: {
        ...data,
      },
    });
  }

  public async delete(id: string) {
    return await this.model.delete({
      where: { id },
    });
  }
}

```

**Comment by ludralph** [comment]
Hi @Lucasnhso 👋 

To address the Capitalization issue, you could use a mapping object that aligns the model names as they are defined in your Prisma schema (typically with an uppercase first letter) with the corresponding lowercase names used by the Prisma client. This mapping could then be used to access the correct model on the PrismaClient instance. For example:
```js
import { PrismaClient } from "@prisma/client";

const prismaClient = new PrismaClient();

// Mapping of Prisma model names from schema (uppercase) to PrismaClient (lowercase)
const modelMapping = {
  User: prismaClient.user,
  Post: prismaClient.post,
  // ... other models
};

export default class BaseRepository {
  protected model;
  
  constructor(modelName: keyof typeof modelMapping) {
    this.model = modelMapping[modelName];
    // ... other method bindings
  }

  // ... repository methods
}
```
The `keyof typeof modelMapping` is used to ensure that only keys that exist in the modelMapping object can be passed as modelName. This approach requires manual upkeep of the mapping object in sync with schema changes.

Alternatively, you could:

1.  Extend Prisma.ModelName with a custom type:
```js
import { PrismaClient, Prisma } from '@prisma/client';

type CustomModelName = string extends Prisma.ModelName ? string : never;
```
2. Use `CustomModelName` in the constructor. `CustomModelName` extends `Prisma.ModelName`, ensuring type safety. It accepts both uppercase and lowercase model names. The model is accessed using `this.prismaClient[modelName.toLowerCase()]`, guaranteeing correct model retrieval regardless of capitalization.
```js
export default class BaseRepository {
  protected model: Prisma.Model<any>;
  protected prismaClient = new PrismaClient();

  constructor(modelName: CustomModelName) {
    this.model = this.prismaClient[modelName.toLowerCase()]; // Ensure lowercase access
    // ... rest of your code
  }
  // ... other methods
}
```
3. You could use it like this:
```js
class UserRepository extends BaseRepository {
  constructor() {
    super('User'); // Pass uppercase model name
  }
}
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by budigital** [comment]
im having the same issue, resolved by using below approach:

```ts
type ModelName = Lowercase<Prisma.ModelName>
```

**Comment by Varenthein** [comment]
I'm probably a little late to the party, but I would suggest something more universal. 

`Lowercase` is enough if you have a one-word model name (like `User` which becomes `user` and so on), but it's not going to work with more detailed names. For instance, `UserSession` becomes `usersession` and it should be `userSession`. So what we really need is not a type that  lowercases the whole name, but rather the first character only. I don't think there is a built-in type for that, but we can easily create one with `infer`.

```
type ToLowerFirstLetter<T extends string> = T extends `${infer F}${infer R}` ? `${Lowercase<F>}${R}` : never;
```

**Reply by mamlzy** [reply]
Thank youuu so much!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
